backend/laundry/views.py

```diff
@@ -152,29 +153,37 @@ class Preferences(APIView):
     """
 
     permission_classes = [IsAuthenticated]
+    key_prefix = "laundry_preferences"
 
     def get(self, request):
+        key = f"{self.key_prefix}:{request.user.id}"
+        cached_preferences = cache.get(key)
+        if cached_preferences is None:
+            preferences = request.user.profile.laundry_preferences.all()
+            cached_preferences = preferences.values_list("hall_id", flat=True)
+            cache.set(key, cached_preferences, MONTH_IN_SECONDS_APPROX)
 
-        preferences = request.user.profile.laundry_preferences.all()
-
-        # returns all hall_ids in a person's preferences
-        return Response({"rooms": preferences.values_list("hall_id", flat=True)})
+        return Response({"rooms": cached_preferences})
 
     def post(self, request):
-
+        key = f"{self.key_prefix}:{request.user.id}"
         profile = request.user.profile
+        preferences = profile.laundry_preferences
 
-        # clears all previous preferences in many-to-many
-        profile.laundry_preferences.clear()
+        if "rooms" not in request.data:
+            return Response({"success": False, "error": "No rooms provided"}, status=400)
 
-        hall_ids = request.data["rooms"]
+        halls = [
+            get_object_or_404(LaundryRoom, hall_id=int(hall_id))
+            for hall_id in request.data["rooms"]
+        ]
 
-        for hall_id in hall_ids:
-            hall = get_object_or_404(LaundryRoom, hall_id=int(hall_id))
-            # adds all of the preferences given by the request
-            profile.laundry_preferences.add(hall)
+        # clears all previous preferences in many-to-many
+        preferences.clear()
+        preferences.add(*halls)
 
-        profile.save()
+        # clear cache
+        cache.delete(key)
```

Comment:
Does this ever error if the key isn't present (which would happen for the first time someone adds a preference)? If so might be worth to check

Reply:
Nope! returns a boolean saying whether or not it actually gets deleted.